akka-actor-tests/src/test/scala/akka/pattern/CircuitBreakerMTSpec.scala

```diff
@@ -36,52 +36,69 @@ class CircuitBreakerMTSpec extends AkkaSpec {
-      val aFewActive = new TestLatch(5)
-      for (_ <- 1 to numberOfTestCalls)
-        yield breaker
-          .withCircuitBreaker(Future {
-            aFewActive.countDown()
-            Await.ready(aFewActive, 5.seconds.dilated)
-            "succeed"
-          })
-          .recoverWith {
-            case _: CircuitBreakerOpenException =>
-              aFewActive.countDown()
-              Future.successful("CBO")
-          }
+    def testCallsWithBreaker(breaker: CircuitBreaker, numberOfCalls: Int): immutable.IndexedSeq[Future[String]] = {
+      for (_ <- 1 to numberOfCalls)
+        yield makeCallWithBreaker(breaker)
     }
 
+    def makeCallWithBreaker(breaker: CircuitBreaker): Future[String] =
+      breaker.withCircuitBreaker(Future.successful("succeed")).recoverWith {
+        case _: CircuitBreakerOpenException =>
+          Future.successful("CBO")
+      }
+
     "allow many calls while in closed state with no errors" in {
-      val futures = testCallsWithBreaker(newBreaker)
+      val futures = testCallsWithBreaker(newBreaker, 10)
       val result = Await.result(Future.sequence(futures), 5.second.dilated)
-      result.size should ===(numberOfTestCalls)
+      result.size should ===(10)
       result.toSet should ===(Set("succeed"))
     }
 
     "transition to open state upon reaching failure limit and fail-fast" in {
       val breaker = newBreaker
       openBreaker(breaker)
-      val futures = testCallsWithBreaker(breaker)
-      val result = Await.result(Future.sequence(futures), 5.second.dilated)
-      result.size should ===(numberOfTestCalls)
-      result.toSet should ===(Set("CBO"))
+
+      breaker.isOpen shouldBe true
+      val call = makeCallWithBreaker(breaker)
+      val result = Await.result(call, 5.second.dilated)
+      result shouldBe "CBO"
     }
 
     "allow a single call through in half-open state" in {
       val breaker = newBreaker
       val halfOpenLatch = new TestLatch(1)
       breaker.onHalfOpen(halfOpenLatch.countDown())
-
       openBreaker(breaker)
 
       // breaker should become half-open after a while
       Await.ready(halfOpenLatch, resetTimeout + 1.seconds.dilated)
 
-      val futures = testCallsWithBreaker(breaker)
-      val result = Await.result(Future.sequence(futures), 5.second.dilated)
-      result.size should ===(numberOfTestCalls)
-      result.toSet should ===(Set("succeed", "CBO"))
+      breaker.isHalfOpen shouldBe true
+
+      val latch = new TestLatch(1)
+      val firstCall =
+        breaker.withCircuitBreaker(Future {
+          // this call closes the CB,
+          // but only after next call fails and touches the latch
+          Await.ready(latch, 5.seconds)
+          "succeed"
+        })
+
+      val secondCall =
+        breaker.withCircuitBreaker(Future.successful("this should have failed")).recoverWith {
+          case _: CircuitBreakerOpenException =>
+            latch.countDown()
+            Future.successful("CBO")
+        }
```

Comment:
More controlled way to confirm that it stays in half-open until the first call completes and closed it.